lib/active_merchant/billing/gateways/sage.rb

```diff
@@ -1,172 +1,422 @@
-      #
-      # ==== Additional options in the +options+ hash for when using a Check as the funding source
-      # * <tt>:originator_id</tt> - 10 digit originator. If not provided, Sage will use the default Originator ID for the specific customer type.
-      # * <tt>:addenda</tt> - Transaction addenda.
-      # * <tt>:ssn</tt> - The customer's Social Security Number.
-      # * <tt>:drivers_license_state</tt> - The customer's drivers license state code.
-      # * <tt>:drivers_license_number</tt> - The customer's drivers license number.
-      # * <tt>:date_of_birth</tt> - The customer's date of birth as a Time or Date object or a string in the format <tt>mm/dd/yyyy</tt>.
-      def purchase(money, source, options = {})
-        if card_brand(source) == "check"
-          virtual_check.purchase(money, source, options)
+        post = {}
+        add_credit_card(post, credit_card)
+        add_transaction_data(post, money, options)
+        commit(:authorization, post, "bankcard")
+      end
+
+      def purchase(money, payment_method, options = {})
+        post = {}
+        if card_brand(payment_method) == "check"
+          source = "virtual_check"
+          add_check(post, payment_method)
+          add_check_customer_data(post, options)
         else
-          bankcard.purchase(money, source, options)
+          source = "bankcard"
+          add_credit_card(post, payment_method)
         end
+        add_transaction_data(post, money, options)
+        commit(:purchase, post, source)
       end
 
-      # Captures authorized funds.
-      #
-      # ==== Parameters
-      #
-      # * <tt>money</tt> - The amount to be authorized as an integer value in cents. Sage doesn't support changing the capture amount, so the full amount of the initial transaction will be captured.
-      # * <tt>reference</tt> - The authorization reference string returned by the original transaction's Response#authorization.
+      # The +money+ amount is not used. The entire amount of the
+      # initial authorization will be captured.
       def capture(money, reference, options = {})
-        bankcard.capture(money, reference, options)
+        post = {}
+        add_reference(post, reference)
+        commit(:capture, post, "bankcard")
       end
 
-      # Voids a prior transaction. Works for both CreditCard and Check transactions.
-      #
-      # ==== Parameters
-      #
-      # * <tt>reference</tt> - The authorization reference string returned by the original transaction's Response#authorization.
       def void(reference, options = {})
-        if reference.split(";").last == "virtual_check"
-          virtual_check.void(reference, options)
-        else
-          bankcard.void(reference, options)
-        end
+        post = {}
+        add_reference(post, reference)
+        source = reference.split(";").last == "virtual_check" ? "virtual_check" : "bankcard"
```

Comment:
I see the string `virtual_check` used in a lot of places - perhaps we could set it as a const?